sdk/kiosk/src/client/kiosk-client.ts

```diff
@@ -0,0 +1,476 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { type SuiClient } from '@mysten/sui.js/src/client';
+import { fetchKiosk, getOwnedKiosks } from '../query/kiosk';
+import {
+	type FetchKioskOptions,
+	KioskData,
+	KioskOwnerCap,
+	ObjectArgument,
+	OwnedKiosks,
+	Network,
+	type KioskClientOptions,
+} from '../types';
+import { TransactionArgument, TransactionBlock } from '@mysten/sui.js/transactions';
+import * as kioskTx from '../tx/kiosk';
+import { TransferPolicyRule, PERSONAL_KIOSK_RULE_ADDRESS, rules } from '../constants';
+import { queryTransferPolicy } from '../query/transfer-policy';
+import { convertToPersonalTx } from '../tx/personal-kiosk';
+import { confirmRequest } from '../tx/transfer-policy';
+import { objArg } from '../utils';
+
+export type PurchaseOptions = {
+	extraArgs?: Record<string, any>;
+};
+
+/**
+ * A Client that allows you to interact with kiosk.
+ * Offers utilities to query kiosk, craft transactions to edit your own kiosk,
+ * purchase, manage transfer policies, create new kiosks etc.
+ */
+export class KioskClient {
+	client: SuiClient;
+	network: Network;
+	#rules: TransferPolicyRule[];
+
+	constructor(options: KioskClientOptions) {
+		this.client = options.client;
+		this.network = options.network;
+		this.#rules = rules; // add all the default rules. WIP
+	}
+
+	// Someone would just have to create a `kiosk-client.ts` file in their project, initialize a KioskClient
+	// and call the `addRuleResolver` function. Each rule has a `resolve` function.
+	// The resolve function is automatically called on `purchaseAndResolve` function call.
+	addRuleResolver(rule: TransferPolicyRule) {
+		if (!this.#rules.find((x) => x.rule === rule.rule)) this.#rules.push(rule);
```

Comment:
I would probably expect either an error or to replace the existing rule here rather than a no-op when an existing rule is found.   If `rule.rule` is a key, it might make more sense to store rules as a map rather than a list

Reply:
Good point, thanks!